tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx

```diff
@@ -0,0 +1,144 @@
+/// \file ROOT/RNTupleIndex.hxx
+/// \ingroup NTuple ROOT7
+/// \author Florine de Geus <florine.de.geus@cern.ch>
+/// \date 2024-04-02
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2024, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RNTupleIndex
+#define ROOT7_RNTupleIndex
+
+#include <ROOT/RField.hxx>
+#include <ROOT/RNTupleUtil.hxx>
+
+#include <unordered_map>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+
+namespace ROOT {
+namespace Experimental {
+namespace Internal {
+/////////////////////////////////////////////////////////////////////////////
+/// Container for the combined hash of the indexed fields. Uses the implementation from `boost::hash_combine` (see
+/// https://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine).
+struct RIndexValue {
+   std::size_t fValue = 0;
```

Comment:
Coming back to this: I think `RIndexValue` now stores the hash of all field values. This will reduce the probability of collisions, but I think it doesn't solve the fundamental problem of finding the "right" entry in case of (a) collision(s). Based on my limited understanding, I think I see two alternatives:
1. Store the values themselves, as Jakob suggests. Then the `std::unordered_map` can (internally) compare the actual values in case of collisions.
2. Alternatively, the `RNTupleIndex` could accept the (rare) occurrence of collisions and use the `std::vector` as candidate entries where either the index or the call site have to (re)load the index column and check if it's actually the value you were looking for.

Reply:
Ah, I realize now I may have misunderstood the original comment. It's a bit tricky since the field values can be of different types, but as you note in your other comment this could be addressed with templating. Perhaps this is indeed something to reconsider...